### PR TITLE
Fon lazy loaded

### PR DESCRIPTION
**Pull request by GiCharkviani**
**Features:**

- FON study in DSM is lazy loaded;
- NgRx store basic implementation and connected to FON study;

**SDK changes:**

- Added forDSM() method for having single instances of specific services in Lazy loaded modules as well, in order to avoid injection errors.

**Interfaces:**
Those files that are located in /models/ directory are just interfaces of existing classes (which classes later will be removed and these interfaces will be used instead), therefore you can skip reviewing them.
 
**Lazy Loaded FON:**
Everything that is located in /FON/ directory belongs to FON lazy-loaded module and they will constantly be improving as we progress: adding new files, developing existing ones and etc.

**STORE:**
Directory / STORE / keeps the files that are mostly related to NgRx Store (Reducers, Actions, Effects, Selectors and etc.).
Store communicates with the rest of the application by its singletone service - store.service.ts- where all necessary function calls are implemented, for example, dispatching actions or returning selectors from the store.

### CODE REVIEW

File: ddp-workspace/projects/ddp-sdk/src/lib/ddp.module.ts [hunk 1]
```diff
@@ -523,4 +523,14 @@ export function createTranslateLoader(http: HttpClient): TranslateHttpLoader {
     ]
 })
 export class DdpModule {
+    static forDSM(): ModuleWithProviders<DdpModule> {
```
Comment (GiCharkviani) [inline ddp-workspace/projects/ddp-sdk/src/lib/ddp.module.ts:526]: used in DSM only

**Review by josemanuel167 — APPROVED** [review]

**Comment by MocanaAtBroad** [comment]
Giorgi: if you want help with reviews you will have help us out with comments. 75 files of Angular application that introduce NgRx and tries to implement lazy loading is lot to process.

No doubt the changes you have made here have a particular strategy and approach in mind. I suggest to use the description field to describe the major things you have done and what decisions you have had to make to make them happen.
And then lots of comments within the files.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/FON/fon.module.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  exports: []
+})
+
+export class fonModule {
+  constructor(private storeService: StoreService,
+              private translateService: TranslateService,
+              private sessionService: SessionService,
+              private dssSessionService: SessionMementoService,
+              private title: Title) {
+
```
Comment (GiCharkviani) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/FON/fon.module.ts:62]: Fon module is used for initializing important configurations, like setting a study name in STORE, translations, DSM token into DSS and etc.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/FON/pages/participantsList/participantsList.component.ts [hunk 3]
```diff
@@ -15,18 +17,23 @@ export class ParticipantsListComponent implements OnInit {
...
 
-  constructor(private agent: AgentService, private router: Router, private activatedRoute: ActivatedRoute) {
+  constructor(private router: Router, private activatedRoute: ActivatedRoute, private storeService: StoreService) {
   }
 
   ngOnInit(): void {
-    this.patients$ = this.agent.getPatients();
-    this.totalCount$ = this.agent.getPatientsTotalCount();
-    this.loading$ = this.agent.isLoading();
+    this.totalCount$ = this.storeService.getParticipantsTotalCount;
```
Comment (GiCharkviani) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/FON/pages/participantsList/participantsList.component.ts:32]: This is how we assign observable selectors to local observable properties in order to observe changes in store, therefore this component always stays informed about the latest changes.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/FON/pages/participantsList/participantsList.component.ts [hunk 4]
```diff
@@ -42,8 +49,8 @@ export class ParticipantsListComponent implements OnInit {
         tap((params: Params) => {
           this.pageSize = params.to - params.from;
           this.pageIndex = (params.to/this.pageSize) - 1;
-        }),
-        switchMap((params: Params) => this.agent.setPage(params.from, params.to))
+          this.patients$ = this.storeService.getParticipants(params.from, params.to, this.PARENT);
```
Comment (GiCharkviani) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/FON/pages/participantsList/participantsList.component.ts:52]: This is an example of dispatching action from Store service in order to trigger changes in the store.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/STORE/actions/enums/participants-enums.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+export enum getParticipants {
+  REQUEST = '[STORE] getParticipants Request',
```
Comment (GiCharkviani) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/STORE/actions/enums/participants-enums.ts:2]: In such files are just written action types (according to the suggested convention) that later are used to distinguish different types of actions from each other.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/STORE/actions/participants.actions.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import {getParticipant, getParticipants} from './enums/participants-enums';
+import {ViewFilter} from '../../filter-column/models/view-filter.model';
+import {Sort} from '../../sort/sort.model';
+import {ParticipantModel} from '../models/participant.model';
+
+export const getParticipantsRequest = createAction(
+  getParticipants.REQUEST,
+  (viewFilter: ViewFilter, realm: string, parent: string, filterQuery: string,
+   from: number, to: number, sort: Sort) => ({viewFilter, realm, parent, filterQuery, from, to, sort})
+);
```
Comment (GiCharkviani) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/STORE/actions/participants.actions.ts:11]: We have 3 different states for actions that also are triggering HTTP requests: Request, Success, and Failure, which is helping us to have control over all expected statuses of specific requests.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/STORE/effects/participants.effects.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    private actions$: Actions,
+    private dsmService: DSMService,
+  ) {
+  }
+
+  getParticipants$ = createEffect(() => this.actions$.pipe(
+      ofType(ParticipantActions.getParticipantsRequest),
+      exhaustMap(({viewFilter, realm, parent, filterQuery, from, to, sort}) =>
+        this.dsmService.applyFilter(viewFilter, realm, parent, filterQuery, from, to, sort)
+          .pipe(
```
Comment (GiCharkviani) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/STORE/effects/participants.effects.ts:20]: This is the file where we control HTTP requests (their frequency, flow, state and etc.).

File: ddp-workspace/projects/ddp-dsm-ui/src/app/STORE/selectors/Utils/participantsList.util.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import {ParticipantModel} from '../../models/participant.model';
+
+export const generateParticipantsList = (participants: ParticipantModel[], settings: any): Array<any> => {
+  const actDefs = settings?.activityDefinitions;
+
+  return participants instanceof Array && actDefs ? participants.map(pt => ({
+    ID: pt.esData.profile.hruid,
```
Comment (GiCharkviani) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/STORE/selectors/Utils/participantsList.util.ts:7]: In the Store, we also have a utils folder, that will help is for calculations that might be needed for selectors which are processing data.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/STORE/store.reducer.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+  error: string | null;
+}
+
+// Initial State
+export const StoreInitialState: StoreStateModel = {
+  settings: {
+    data: {},
+    isLoaded: false
+  },
+  participants: {
```
Comment (GiCharkviani) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/STORE/store.reducer.ts:24]: This is the file where we communicate with the store.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/app.module.ts [hunk 10]
```diff
@@ -58,81 +45,32 @@ sdkConfig.dashboardPageUrl = '/fon/patients';
...
   ],
   imports: [
     BrowserModule,
+    AppRoutingModule,
     BrowserAnimationsModule,
-    DdpModule,
     HttpClientModule,
-    AppRoutingModule,
-    ...material
+    StoreModule.forRoot({MainStore: storeReducer}),
```
Comment (GiCharkviani) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/app.module.ts:62]: At this line, we can add new stores as well in the future.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/FON/pages/participantsList/participantsList.component.html [hunk 11]
```diff
@@ -1,5 +1,6 @@
 <div class="participants">
   <h3>Patients list</h3>
+  <h2 style="color: red">{{ error$ | async }}</h2>
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/FON/pages/participantsList/participantsList.component.html:3]: **_#minor_**

TODO: fix the styles

Comment (GiCharkviani) [reply ddp-workspace/projects/ddp-dsm-ui/src/app/FON/pages/participantsList/participantsList.component.html:3]: Of course. It's very very temporary what you just saw :D :D :D . It was only for testing purposes

**Comment by alextanlan** [comment]
We agreed to merge the PR right now and check it more carefully later again.

**Review by alextanlan — APPROVED** [review]

**Pull request merged**